Documentation on tensor parallelism (#339)
huggingface/optimum-neuron: docs/source/guides/distributed_training.mdx

```python
from torch.optim import AdamW
from optimum.neuron.distributed import make_optimizer_constructor_lazy

lazy_adamw = make_optimizer_constructor_lazy(AdamW)
```

  2. Set the `zero_1` argument to `True` when instantiating the `NeuronAccelerator`.

```python
accelerator = NeuronAccelerator(
    ...
    zero_1=True,
)
```

## How to enable Tensor Parallelism?

Just as for ZeRO-1, it is possible to use Tensor Parallelism both with the [`NeuronTrainer`] and the [`NeuronAccelerator`].

When doing Tensor Parallelism, you have different settings:
  1. The `tensor_parallel_size`. Ideally it should be smallest value for which the model fits. 
  2. Whether or not sequence parallelism should be enabled. [Sequence parallelism](https://arxiv.org/pdf/2205.05198.pdf) shards the activations on the sequence axis outside of the tensor parallel regions.
  It is useful because it saves memory by sharding the activations.
  3. Whether or not parallelization of the embedding (and thus the LM head for decoder / seq2seq models) should be done. **It is not supported yet**.

> On top of that, it is very important to make sure that the original model is loaded in an efficient manner: the training script is going to be called by `torchrun`, which will dispatch it to workers. If each worker (there are 32 of them in  a `trn1.32xlarge` instance) loads the full model weights, it can take a lot of time and go out-of-memory really fast.

Reviewer: Explain that one must allocate one worker per core you want to use.

> `optimum-neuron` provides a context-manager [`distributed.lazy_load_for_parallelism`] that loads the model lazily to prevent that, only the sharded model will be materialized in each worker.

Reviewer: ```suggestion
`optimum-neuron` provides a context-manager [`distributed.lazy_load_for_parallelism`] that loads the model lazily to prevent that, only the parameters for the corresponding model shard will be materialized in each worker.
```

> <!--Copyright 2023 The HuggingFace Team. All rights reserved.

Reviewer 2: Could you also add it to the overview of how-to-guides?

the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->
# Distributed Training with `optimum-neuron`

[AWS Trainium instances](https://aws.amazon.com/machine-learning/trainium/) are great to train models. They can contain up to 16 Neuron devices, each device containing 2 Neuron cores and has 32GB of memory (16GB per core). For example a `trn1.32xlarge` instance has 32 x 16 = 512GB of memory.

But there is a caveat: each Neuron core is an independent data-parallel worker by default. It means that the model, the gradient state and the optimizer state, amounting to approximately 4 times the model size, must fit in each of the Neuron cores (16GB) to be able to train. If that is the case, then the activations must also fit in the remaining memory.

> To alleviate that, `optimum-neuron` supports parallelism features enabling your to harness the full power of your Trainium instance:

Reviewer 2: ```suggestion
To alleviate that, `optimum-neuron` supports parallelism features enabling you to harness the full power of your Trainium instance:
```

>   1. [ZeRO-1](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/zero1_gpt2.html): It is a optimization of data-parallism which consists in sharding the optimizer state (which usually represents half of the memory needed on the device) over the data-parallel ranks.

Reviewer 2: ```suggestion
  1. [ZeRO-1](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/tutorials/training/zero1_gpt2.html): It is an optimization of data-parallelism which consists in sharding the optimizer state (which usually represents half of the memory needed on the device) over the data-parallel ranks.
```

  2. [Tensor Parallelism](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/neuronx-distributed/tensor_parallelism_overview.html): It is a technique which consists in sharding each of your model parameters along a given dimension on multiple devices. The number of devices to shard your parameters on is called the `tensor_parallel_size`. 
  3. [Pipeline Parallelism](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/neuronx-distributed/pipeline_parallelism_overview.html): **coming soon!**

The good news is that is it possible to combine those techniques, and `optimum-neuron` makes it very easy!

<Tip>

> All the [example scripts](https://github.com/huggingface/optimum-neuron/tree/main/examples) provided in the `optimum-neuron` repo have those features implemented via the [`NeuronTrainer`].

Reviewer 2: Still can't correctly display the link?

Author: Just removed the links in the <Tip> section.

### Via the `NeuronTrainer`

```python
from optimum.neuron import NeuronTrainingArguments, NeuronTrainer

# To enable ZeRO-1, set the `zero_1` argument to `True` in the training arguments.
training_args = NeuronTrainingArguments(
    ...,
    zero_1=True,
)

trainer = NeuronTrainer(
    model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
)

trainer.train()
```

<Tip>
Since the [example scripts](https://github.com/huggingface/optimum-neuron/tree/main/examples) use the [`NeuronTrainer`], you can enable ZeRO-1 when using them by add the `--zero_1` flag
to your command line.

For example:

```bash
torchrun --nproc_per_node=2 examples/language-modeling/run_clm.py \
    --model_name_or_path TinyLlama/TinyLlama-1.1B-Chat-v0.6 \
    --dataset_name wikitext \
    --dataset_config_name wikitext-2-raw-v1 \
    --do_train \
    --per_device_train_batch_size 1 \
    --block_size 1024 \
    --bf16 \
    --zero_1 \
    --output_dir my_training/

```
</Tip>

### Via the `NeuronAccelerator`

There is a little bit more work to do when not using the `NeuronTrainer`:
  

>   1. (Optional) Wrap the optimizer class to make it lazy. When ZeRO-1 is enabled the original optimizer is overridden to use a sharded version of it so it is possible to load original optimizer lazily so that the optimizer state is not materialized until it is actually sharded.

Reviewer 2: ```suggestion
  1. (Optional) Wrap the optimizer class to make it lazy. When ZeRO-1 is enabled the original optimizer is overridden to use a sharded version of it. Hence, it is possible to load the original optimizer lazily so that the optimizer state is not materialized until it is actually sharded.
```

```python
from torch.optim import AdamW
from optimum.neuron.distributed import make_optimizer_constructor_lazy

lazy_adamw = make_optimizer_constructor_lazy(AdamW)
```

  2. Set the `zero_1` argument to `True` when instantiating the `NeuronAccelerator`.

```python
accelerator = NeuronAccelerator(
    ...
    zero_1=True,
)

model = ...
lazy_optimizer = lazy_adamw(...) # Actually instantiate the optimizer.

model, optimizer = accelerator.prepare(model, lazy_optimizer)
```

## How to enable Tensor Parallelism?

> Just as for ZeRO-1, it is possible to use Tensor Parallelism both with the [`NeuronTrainer`] and the [`NeuronAccelerator`].

Reviewer 2: ```suggestion
Just as for ZeRO-1, it is possible to apply Tensor Parallelism either with the [`NeuronTrainer`] or the [`NeuronAccelerator`].
```
